### PR TITLE
Gpu ordered_logistic_glm_lpmf and categorical_logit_glm_lpmf

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Adds OpenCL implementations for `ordered_logistic_glm_lpmf()` and `categorical_logit_glm_lpmf()`. Two new OpenCL device functions (`log1p_exp()` and `log1m_exp()`) as they are needed in `ordered_logistic_glm_lpmf()`. They are essentialy the same as existing CPU functions with same name in stan math.

## Tests
There are new tests for both OpenCL implementations of GLMs  and both device functions.

## Side Effects

None.

## Checklist

- [x] Math issue #1319

- [x] Copyright holder: Tadej Ciglarič (University of Ljubljana)

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
I will review this one tomorrow.

File: stan/math/opencl/device_functions/log1m_exp.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#ifndef STAN_MATH_OPENCL_DEVICE_FUNCTIONS_LOG1M_EXP_HPP
+#define STAN_MATH_OPENCL_DEVICE_FUNCTIONS_LOG1M_EXP_HPP
```
Comment (rok-cesnovar) [inline stan/math/opencl/device_functions/log1m_exp.hpp]: I missed this one on the digamma case, but can we move the device functions inside kernels. They are only going to be used inside kernels, so I think that makes most sense.

File: stan/math/opencl/device_functions/log1p_exp.hpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#ifndef STAN_MATH_OPENCL_DEVICE_FUNCTIONS_LOG1P_EXP_HPP
+#define STAN_MATH_OPENCL_DEVICE_FUNCTIONS_LOG1P_EXP_HPP
```
Comment (rok-cesnovar) [inline stan/math/opencl/device_functions/log1p_exp.hpp]: Same as for log1m location.

File: stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp [hunk 3]
```diff
@@ -64,7 +64,7 @@ categorical_logit_glm_lpmf(
   check_bounded(function, "categorical outcome out of support", y, 1,
                 N_classes);
 
-  if (size_zero(y, x, beta)) {
+  if (size_zero(y, x, beta) || N_classes == 1) {
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp]: Is thus a bug in the CPU version you are addressing here or just a performance thing for this case?

File: stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp [hunk 4]
```diff
@@ -75,8 +75,12 @@ ordered_logistic_glm_lpmf(
...
   check_bounded(function, "Vector of dependent variables", y, 1, N_classes);
   check_ordered(function, "Cut-points", cuts);
-  check_finite(function, "Final cut-point", cuts[N_classes - 2]);
-  check_finite(function, "First cut-point", cuts[0]);
+  if (N_classes > 1) {
+    if (N_classes > 2) {
+      check_finite(function, "Final cut-point", cuts[N_classes - 2]);
+    }
+    check_finite(function, "First cut-point", cuts[0]);
+  }
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp]: Can you elaborate on this change. I dont think its mentioned in the PR text.

File: test/unit/math/opencl/device_functions/log1m_exp_test.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+TEST(MathMatrixCL, log1m_exp) {
+  Eigen::VectorXd a = Eigen::VectorXd::Random(1000).array() * 0.9999;
+
+  stan::math::matrix_cl<double> a_cl(a);
+  stan::math::matrix_cl<double> res_cl(1000, 1);
+  log1m_exp(cl::NDRange(1000), res_cl, a_cl);
+  Eigen::VectorXd res = stan::math::from_matrix_cl<-1, 1>(res_cl);
+
+  stan::test::expect_near_rel("log1m_exp (OpenCL)", res,
+                              stan::math::log1m_exp(a));
```
Comment (rok-cesnovar) [inline test/unit/math/opencl/device_functions/log1m_exp_test.cpp]: Edge case tests are needed. See https://github.com/stan-dev/math/blob/502487c511594ccb93eb979df5b8fe163becb417/test/unit/math/prim/scal/fun/log1m_exp_test.cpp

File: test/unit/math/opencl/device_functions/log1p_exp_test.cpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+TEST(MathMatrixCL, log1p_exp) {
+  Eigen::VectorXd a = Eigen::VectorXd::Random(1000).array() * 0.9999;
+
+  stan::math::matrix_cl<double> a_cl(a);
+  stan::math::matrix_cl<double> res_cl(1000, 1);
+  log1p_exp(cl::NDRange(1000), res_cl, a_cl);
+  Eigen::VectorXd res = stan::math::from_matrix_cl<-1, 1>(res_cl);
+
+  stan::test::expect_near_rel("log1p_exp (OpenCL)", res,
```
Comment (rok-cesnovar) [inline test/unit/math/opencl/device_functions/log1p_exp_test.cpp]: Edge cases need to be covered: https://github.com/stan-dev/math/blob/502487c511594ccb93eb979df5b8fe163becb417/test/unit/math/prim/scal/fun/log1p_exp_test.cpp

File: stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp (hunk 3 above)
Comment (t4c1) [reply stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp]: A bit of both and a bit of simplifying testing. The problem happend with testing this case. Result was correctly calculated as (very close to) 0 in both CPU and GPU impleemntation. However, there were a bit different numerical errors (absolutely tiny, but relatively large) that caused `expect_near_rel()` to fail the test.

File: stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp (hunk 4 above)
Comment (t4c1) [reply stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp]: This was a bug. Thise checks could cause segfault if number of classes was less than 2.

File: stan/math/opencl/device_functions/log1m_exp.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/opencl/device_functions/log1m_exp.hpp]: I am also moving digamma in this PR.

Comment (rok-cesnovar) [reply stan/math/opencl/device_functions/log1m_exp.hpp]: Yes, please do.

**Pull request merged**